File: 0001~0500/0169. Majority Element/main.cpp
/** 
 * Source: https://leetcode.com/problems/majority-element/
 * Date: 2022/11/30
 * Skill: 
 * Runtime: 11 ms, faster than 99.57%
 * Memory Usage: 19.6 MB, less than 15.21%
 * Time complexity: 
 * Space complexity: 
 * Constraints: 
 *      
 */

#include <iostream>
#include <vector>
#include <algorithm> 
using namespace std;

class Solution {
public:
    int majorityElement(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        return nums[nums.size()/2];
    }
};

int main() {
    
}

static const auto io_sync_off = []() {
    // turn off sync
    std::ios::sync_with_stdio(false);
    // untie in/out streams
    std::cin.tie(nullptr);
    return nullptr;
}();

File: 0001~0500/0169. Majority Element/main.py
"""
Source: https://leetcode.com/problems/majority-element/
Date: 2022/11/30
Skill: Divide and Conquer
Runtime: 259 ms, faster than 71.89%
Memory Usage: 15.9 MB, less than 35.29%
Time complexity: 
Space complexity: 
Constraints: 
    
"""

from typing import List

class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        def majority_element_rec(lo,hi) -> int:
            if lo == hi:
                return nums[lo]
            mid = (lo+hi)//2
            left = majority_element_rec(lo,mid)
            right = majority_element_rec(mid+1,hi)
            if left == right:
                return left
            left_count = sum(1 for i in range(lo, mid+1) if nums[i] == left)
            right_count = sum(1 for i in range(mid+1, hi+1) if nums[i] == right)
            if left_count > right_count:
                return left
            else:
                return right
            
        return majority_element_rec(0,len(nums)-1)